index.bs

```diff
@@ -1174,64 +1174,92 @@ The {{MLOpSupportLimits}} has the following top level members, aside from these,
 <script type="idl">
 dictionary MLOpSupportLimits {
   MLInputOperandLayout preferredInputLayout;
-  MLSupportLimits input;
-  MLSupportLimits constant;
-  MLSupportLimits output;
+  [EnforceRange] unsigned long long maxTensorByteLength;
+  MLDataTypeLimits input;
+  MLDataTypeLimits constant;
+  MLDataTypeLimits output;
 };
 </script>
 
 <dl dfn-type=dict-member dfn-for=MLOpSupportLimits>
     : <dfn>preferredInputLayout</dfn>
     :: Preferred input layout for layout dependent operators like {{MLGraphBuilder/conv2d()}}.
+    : <dfn>maxTensorByteLength</dfn>
+    :: The maximum supported length of tensors, in bytes.
     : <dfn>input</dfn>
     :: Support limits for input {{MLOperand}}s for an {{MLGraph}}.
     : <dfn>constant</dfn>
     :: Support limits for constant {{MLOperand}}s for an {{MLGraph}}.
     : <dfn>output</dfn>
     :: Support limits for output {{MLOperand}}s for an {{MLGraph}}.
 </dl>
-#### {{MLSupportLimits}} dictionary #### {#api-mlcontext-supportlimits-dictionary}
+#### {{MLDataTypeLimits}} dictionary #### {#api-mlcontext-datatypelimits-dictionary}
 <script type="idl">
-dictionary MLSupportLimits {
+dictionary MLDataTypeLimits {
   sequence<MLOperandDataType> dataTypes;
 };
 </script>
-<dl dfn-type=dict-member dfn-for=MLSupportLimits>
+<dl dfn-type=dict-member dfn-for=MLDataTypeLimits>
     : <dfn>dataTypes</dfn>
     :: Supported data types.
 </dl>
 
+#### {{MLRankRange}} dictionary #### {#api-mlcontext-rankrange-dictionary}
+<script type="idl">
+dictionary MLRankRange {
+  unsigned long min;
+  unsigned long max;
+};
+</script>
+<dl dfn-type=dict-member dfn-for=MLRankRange>
+    : <dfn>min</dfn>
+    :: Minimum supported rank.
+    : <dfn>max</dfn>
+    :: Maximum supported rank.
+</dl>
+
+#### {{MLTensorLimits}} dictionary #### {#api-mlcontext-tensorlimits-dictionary}
+<script type="idl">
+dictionary MLTensorLimits : MLDataTypeLimits {
+  MLRankRange rankRange;
+};
+</script>
+<dl dfn-type=dict-member dfn-for=MLTensorLimits>
+    : <dfn>rankRange</dfn>
+    :: Minimum and maximum supported ranks.
+</dl>
+
 #### {{MLBinarySupportLimits}} dictionary #### {#api-mlcontext-binarysupportlimits-dictionary}
 <script type="idl">
 dictionary MLBinarySupportLimits {
-  MLSupportLimits a;
-  MLSupportLimits b;
-  MLSupportLimits output;
+  MLTensorLimits a;
+  MLTensorLimits b;
+  MLDataTypeLimits output;
```

Comment:
Isn't the output rank also informative?

```javascript
  MLTensorLimits output;
```

If we say that the reason the output rank *isn't useful* is because the operator will always produce a valid output rank given valid inputs (and thus it doesn't matter anyway), then should we also presume that an operator will always produce a valid output data type (barring `cast` and `argMin`/`argMax`) given valid inputs, and thus the output data type is not necessary either? I'm not actually arguing to remove `output` data type, but rather pointing out the odd asymmetry, which justifies output rank's existence. If output rank matters for no other operator, it at least matters for `reshape` (where the output shape is not inferred from the input tensors).

Reply:
> it at least matters for `reshape` (where the output shape is not inferred from the input tensors).

+1, also `expand` whose output rank is defined the size of `newShape`. Reduction ops may also be concerned, say if the platform doesn't support scalar output (rank 0), then reduce all dimensions with `keepDimensions = false` would not be supported.

Reply:
Thoughts @reillyeon ?

Reply:
I agree that since the output type and shape are always a function of the inputs and options we don't need to specify what's supported here explicitly. This does mean that we need a different way of expressing constraints like the data types that can be passed to `cast()` which can be implied by the supported output data types.

Reply:
How about:
* Removing `output` members from all of the support limits dictionaries.
* Adding this dictionary type, which would be used for `cast()`, `argMin()`, `argMax()`

```webidl
dictionary MLSingleInputWithOutputDataTypeSupportLimits {
  MLTensorLimits input;
  sequence<MLOperandDataType> outputDataTypes;
}
```

Reply:
`argMin()` and `argMax()` call this argument `outputDataType` while `cast()` calls it `type`. I'd rather have the `opSupportLimits()` structure use the argument names. Could we rename both arguments to `dataType`?

Reply:
Re: renaming arguments - `outputDataType` in the dicts for `argMin()`/`argMax()` would be a breaking change (is ORT using this yet?) but that shouldn't hold us back right now. But I personally like the self-documenting nature of including "output" for someone reading a code snippet; if it were just `dataType` the reader might wonder why that wasn't implicit. Changing `cast()`'s argument name is a non-breaking change since it's not web-visible. 

Re: the extremes of either including ranks in `output`, or only including output data types when it's an input argument - any further opinions? I lean towards keeping the API... lean. But I acknowledge I'm a weirdo.

Reply:
I agree with keeping the API lean and just documenting valid input parameters.

Reply:
> Re: the extremes of either including ranks in `output`, or only including output data types when it's an input argument - any further opinions?

I slightly prefer to have ranks in `output`. It basically represents output row in each op's constraints table.

Reply:
Added #835 to capture some of this discussion.

Reply:
I also prefer having ranks for output, the `reshape` example that @fdwr bring up is a valid one that makes [CoreML backend output rank validation](https://source.corp.google.com/h/webium-private/prototype/+/main:services/webnn/coreml/graph_builder_coreml.cc;l=5144;drc=a6b7e04eeeef31a788fd9188dbb76de8d37ee708) return an async error instead of type error right now.

Reply:
Resolving this comment which is now captured by: https://github.com/webmachinelearning/webnn/issues/835